### PR TITLE
HF-85: Implement all database functions

### PR DESCRIPTION
**Pull request by marcin-kordas-hoc**
### Context

Implements all 12 Excel database functions (D-functions family). Originally scoped to DCOUNT only, expanded to the full family since all share the same infrastructure (field resolution, criteria parsing, row matching).

### How did you test your changes?

- 185 unit tests in hyperformula-tests (handsontable/hyperformula-tests#9)
- 167-case Excel validation workbook (all PASS in Excel Desktop)
- 147-test runtime integration suite + 30 edge case tests (booleans, negatives, zeros, wildcards, large DB, comparison operators)
- Verified all error types match Excel precisely (#VALUE!, #DIV/0!, #NUM!)

### Types of changes

- [x] New feature or improvement (a non-breaking change that adds functionality)
- [x] Additional language file, or a change to an existing language file (translations)
- [x] Change to the documentation

### Related issues:
1. Fixes HF-85

### Checklist:

- [x] I have reviewed the guidelines about Contributing to HyperFormula and I confirm that my code follows the code style of this project.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [x] I described my changes in the CHANGELOG.md file.
- [x] My changes require a documentation update.

---

## Summary

- 12 database functions: DCOUNT, DCOUNTA, DSUM, DAVERAGE, DMAX, DMIN, DGET, DPRODUCT, DSTDEV, DSTDEVP, DVAR, DVARP
- New `DatabasePlugin` (533 lines) with shared infrastructure
- i18n translations for all 17 languages (proper Excel-localized names)
- Documentation: `built-in-functions.md` (Database section), `known-limitations.md` (Nuances)

## Implementation

- `withDatabaseArgs()` helper eliminates boilerplate across all 12 functions
- `resolveFieldIndex()` — string (case-insensitive header match) or 1-based numeric index with `Math.trunc()`
- `buildDatabaseCriteria()` — OR across rows, AND within row, reuses `CriterionBuilder`
- `rowMatchesCriteria()` — `.some()` (OR) + `.every()` (AND)
- `collectNumericValues()` — shared by DSTDEV/DSTDEVP/DVAR/DVARP

## Excel behavior edge cases

| Function | Edge case | Behavior |
|---|---|---|
| DMAX, DMIN, DPRODUCT | No matches | Returns 0 |
| DGET | 0 matches / 2+ matches | #VALUE! / #NUM! |
| DAVERAGE | No numeric values | #DIV/0! |
| DSTDEV, DVAR | ≤1 value | #DIV/0! (sample, n-1) |
| DSTDEVP, DVARP | 1 value / 0 values | 0 / #DIV/0! (population, n) |

## Linked

- Tests PR: handsontable/hyperformula-tests#9
- ClickUp: HF-85

---

> [!NOTE]
> **Medium Risk**
> Adds new interpreter functionality that affects formula evaluation semantics (criteria parsing, error handling, and aggregation behavior), though changes are largely additive and isolated to a new plugin plus docs/i18n updates.
> 
> **Overview**
> Adds a new `DatabasePlugin` implementing the 12 Excel database functions (`DCOUNT`, `DCOUNTA`, `DSUM`, `DAVERAGE`, `DMAX`, `DMIN`, `DGET`, `DPRODUCT`, `DSTDEV`, `DSTDEVP`, `DVAR`, `DVARP`), including shared logic for field resolution, criteria parsing, row matching, and Excel-like error propagation.
> 
> Updates the public surface by exporting the plugin, adding translations for these functions across language packs, and expanding docs/CHANGELOG to include a new **Database** functions category and function list.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 74c43972528c85c5594f5268af11734f8a08e6cd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [HF-85 Implement function DCOUNT](https://app.clickup.com/t/9015210959/HF-85)

File: src/i18n/languages/csCZ.ts [hunk 1]
```diff
@@ -75,6 +75,18 @@ const dictionary: RawTranslationPackage = {
     DAYS360: 'ROK360',
     DAYS: 'DAYS',
     DB: 'ODPIS.ZRYCH',
+    DAVERAGE: 'DPRUMER',
+    DCOUNT: 'DCOUNT',
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/i18n/languages/csCZ.ts]: Fixed in commit 594c9a26c. Changed `DCOUNT: 'DCOUNT'` to `DCOUNT: 'DPOCET'` in csCZ.ts, consistent with DCOUNTA = 'DPOCET2'.

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+    if (isExtendedNumber(field)) {
+      const index = Math.trunc(getRawValue(field))
+      if (index < 1 || index > headers.length) {
+        return new CellError(ErrorType.VALUE, ErrorMessage.WrongType)
+      }
+      return index - 1
+    }
+
+    return new CellError(ErrorType.VALUE, ErrorMessage.WrongType)
+  }
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:490]: Fixed in commit 594c9a26c. `resolveFieldIndex` now coerces boolean field arg to number (TRUE → 1, FALSE → 0), matching HF's ArithmeticHelper convention and Excel semantics. Regression test added in hyperformula-tests#9 (commit 7c43a8f).

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+      return new CellError(ErrorType.VALUE, ErrorMessage.WrongType)
+    }
+
+    const numericField = typeof field === 'boolean' ? Number(field) : field
+
+    if (isExtendedNumber(numericField)) {
+      const index = Math.trunc(getRawValue(numericField))
+      if (index < 1 || index > headers.length) {
+        return new CellError(ErrorType.VALUE, ErrorMessage.WrongType)
+      }
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:485]: Fixed in commit 11b0488da. Replaced the bounds check with `!Number.isFinite(index) || index < 1 || index > headers.length` — catches NaN and ±Infinity defensively. Note: in practice HF propagates NaN-producing expressions (LOG(-1), 1/"a") as CellError before reaching resolveFieldIndex, so this guard is defensive for future call paths. Regression test added in hyperformula-tests#9 (commit de2eef5) documenting the error-propagation behavior.

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+    if (isExtendedNumber(numericField)) {
+      const index = Math.trunc(getRawValue(numericField))
+      if (!Number.isFinite(index) || index < 1 || index > headers.length) {
+        return new CellError(ErrorType.VALUE, ErrorMessage.WrongType)
+      }
+      return index - 1
+    }
+
+    return new CellError(ErrorType.VALUE, ErrorMessage.WrongType)
+  }
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:490]: Fixed in commit 3b238de92. Both `resolveFieldIndex` and `buildDatabaseCriteria` now check `instanceof CellError` upfront and return the error verbatim, preserving the original type (#DIV/0!, #NUM!, etc.) instead of masking it as #VALUE! / BadCriterion. Regression tests added in hyperformula-tests#9 (commit 46f53f0) covering both the field-argument and criteria-cell paths.

File: src/i18n/languages/nlNL.ts [hunk 5]
```diff
@@ -75,6 +75,18 @@ const dictionary: RawTranslationPackage = {
     DAYS360: 'DAGEN360',
     DAYS: 'DAGEN',
     DB: 'DB',
+    DAVERAGE: 'DBGEMIDDELDE',
+    DCOUNT: 'DBAANTAL',
+    DCOUNTA: 'DBAAANTALC',
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/i18n/languages/nlNL.ts]: Fixed in commit 0099900c5. `DCOUNTA: 'DBAAANTALC'` → `DCOUNTA: 'DBAANTALC'` (matches DB + AANTAL + C pattern, consistent with DCOUNT = DBAANTAL).

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+   */
+  public dsum(ast: ProcedureAst, state: InterpreterState): InterpreterValue {
+    return this.withDatabaseArgs(ast, state, 'DSUM', (dbData, fieldIndex, criteriaRows) => {
+      let sum = 0
+
+      for (let rowIdx = 1; rowIdx < dbData.length; rowIdx++) {
+        if (this.rowMatchesCriteria(dbData[rowIdx], criteriaRows)) {
+          const cellValue = dbData[rowIdx][fieldIndex]
+          if (isExtendedNumber(cellValue)) {
+            sum += getRawValue(cellValue)
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:184]: Fixed in commit 947a9fdbf. Each numeric aggregation now returns a CellError from a matching field cell immediately instead of letting `isExtendedNumber` silently skip it. `collectNumericValues` (shared by DSTDEV/DSTDEVP/DVAR/DVARP) returns `CellError | number[]`, so all four statistical functions propagate too. DCOUNT/DCOUNTA/DGET left unchanged — COUNT-family Excel semantics skip errors, and DGET already returns the cell verbatim. Regression test added in hyperformula-tests#9 (commit 6fbd56c).

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 7]
```diff
@@ -0,0 +1,576 @@
...
+    return this.withDatabaseArgs(ast, state, 'DCOUNT', (dbData, fieldIndex, criteriaRows) => {
+      let count = 0
+
+      for (let rowIdx = 1; rowIdx < dbData.length; rowIdx++) {
+        if (this.rowMatchesCriteria(dbData[rowIdx], criteriaRows)) {
+          if (isExtendedNumber(dbData[rowIdx][fieldIndex])) {
+            count++
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:107]: Fixed in commit 2727f8eae. DCOUNT now also propagates CellError from matching field cells, matching the other 9 numeric D-functions. Regression test added in hyperformula-tests#9 (commit 3f986ed).

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+      let count = 0
+
+      for (let rowIdx = 1; rowIdx < dbData.length; rowIdx++) {
+        if (this.rowMatchesCriteria(dbData[rowIdx], criteriaRows)) {
+          const cellValue = dbData[rowIdx][fieldIndex]
+          if (cellValue !== EmptyValue && cellValue !== undefined && cellValue !== null) {
+            count++
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:133]: Fixed in commit 2727f8eae. DCOUNTA now also propagates CellError from matching field cells for consistency with the other D-functions. Regression test added in hyperformula-tests#9 (commit 3f986ed).

File: src/i18n/languages/plPL.ts [hunk 9]
```diff
@@ -75,6 +75,18 @@ const dictionary: RawTranslationPackage = {
     DAYS360: 'DNI.360',
     DAYS: 'DNI',
     DB: 'DB',
+    DAVERAGE: 'BD.SREDNIA',
+    DCOUNT: 'BD.ILE.REKORDÓW',
+    DCOUNTA: 'BD.ILE.REKORDOW.A',
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/i18n/languages/plPL.ts]: Fixed in commit 4e7354550. `DCOUNTA: 'BD.ILE.REKORDOW.A'` → `DCOUNTA: 'BD.ILE.REKORDÓW.A'` (verified against Microsoft Polish function docs). Also caught a similar missing Ś in the same file — DAVERAGE `'BD.SREDNIA'` → `'BD.ŚREDNIA'`. Both now use proper Polish diacritics.

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+    return this.withDatabaseArgs(ast, state, 'DGET', (dbData, fieldIndex, criteriaRows) => {
+      let matchedValue: InternalScalarValue | undefined
+      let matchCount = 0
+
+      for (let rowIdx = 1; rowIdx < dbData.length; rowIdx++) {
+        if (this.rowMatchesCriteria(dbData[rowIdx], criteriaRows)) {
+          const cellValue = dbData[rowIdx][fieldIndex]
+          if (cellValue instanceof CellError) {
+            return cellValue
+          }
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts]: Fixed in commit 9d07107d7. The CellError check is now after the loop, so the multi-match #NUM! takes priority — a field error is only propagated when exactly one row matches. The earlier defensive guard (from 4e7354550) was the regression cause, this commit reverts it to post-loop handling. Two regression tests added in hyperformula-tests#9 (commit 0110ca2) — one for the multi-match-with-first-error case, one for the single-match-with-error case.

File: docs/guide/integration-with-angular.md [hunk 11]
```diff
@@ -4,6 +4,70 @@ Installing HyperFormula in an Angular application works the same as with vanilla
 
 For more details, see the [client-side installation](client-side-installation.md) section.
 
+## Basic usage
```
Comment (sequba) [inline docs/guide/integration-with-angular.md]: The framework integration guides should not be modified in this PR

File: docs/guide/known-limitations.md [hunk 12]
```diff
@@ -38,3 +38,5 @@ you can't compare the arguments in a formula like this:
 * The INDEX function doesn't support returning whole rows or columns of the source range – it always returns the contents of a single cell.
 * The FILTER function accepts either single rows of equal width or single columns of equal height. In other words, all arrays passed to the FILTER function must have equal dimensions, and at least one of those dimensions must be 1.
 * Array-producing functions (e.g., SEQUENCE, FILTER) require their output dimensions to be determinable at parse time. Passing cell references or formulas as dimension arguments (e.g., `=SEQUENCE(A1)`) results in a `#VALUE!` error, because the output size cannot be resolved before evaluation.
+* DGET returns `#VALUE!` when no records match the criteria, and `#NUM!` when more than one record matches.
+* DMAX, DMIN, and DPRODUCT return `0` when no records match the criteria.
```
Comment (sequba) [inline docs/guide/known-limitations.md]: why is it listed as limitations? Is it different to what Excel does?

**Review by sequba — CHANGES_REQUESTED** [review]

File: docs/guide/integration-with-angular.md (hunk 11 above)
Comment (marcin-kordas-hoc) [reply docs/guide/integration-with-angular.md]: Reverted in commit `0c3decd04` — framework integration guide changes were out of scope for HF-85 (D-functions). The corresponding HF-122 PR #1653 ships the actual framework guides.

File: docs/guide/known-limitations.md (hunk 12 above)
Comment (marcin-kordas-hoc) [reply docs/guide/known-limitations.md]: Agreed — DGET / DMAX / DMIN / DPRODUCT behaviors match Excel exactly (DGET returns `#VALUE!` for no match, `#NUM!` for multiple matches; DMAX/DMIN/DPRODUCT return 0 when no match). They are not HyperFormula deviations and do not belong in `known-limitations.md`. Removed in commit `465321fd7`.

File: src/i18n/languages/fiFI.ts [hunk 13]
```diff
@@ -75,6 +75,18 @@ const dictionary: RawTranslationPackage = {
     DAYS360: 'PÄIVÄT360',
     DAYS: 'PV',
     DB: 'DB',
+    DAVERAGE: 'TKESKIARVO',
+    DCOUNT: 'TLASKE',
+    DCOUNTA: 'TLASKE.A',
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/i18n/languages/fiFI.ts:80]: False positive — Finnish Excel uses dot syntax in function names (e.g., `LASKE.A` for COUNTA, `KESKIARVO.JOS` for AVERAGEIF, `LASKE.TYHJÄT` for COUNTBLANK). `TLASKE.A` for DCOUNTA is consistent with this convention and matches Microsoft's official Finnish translation. The dot is intentional, not a separator typo.

File: src/i18n/languages/ruRU.ts [hunk 14]
```diff
@@ -75,6 +75,18 @@ const dictionary: RawTranslationPackage = {
     DAYS360: 'ДНЕЙ360',
     DAYS: 'ДНИ',
     DB: 'ФУО',
+    DAVERAGE: 'БДСРЕДНЕЕ',
+    DCOUNT: 'БСЧЁТ',
+    DCOUNTA: 'БСЧЕТА',
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/i18n/languages/ruRU.ts]: Fixed in commit `fae8844f1` — `БСЧЕТА` → `БСЧЁТА`. Verified against the COUNT/COUNTA/DCOUNT translations in the same file (all use Ё) and Microsoft's official Russian Excel translation.

File: src/interpreter/plugin/DatabasePlugin.ts [hunk 15]
```diff
@@ -0,0 +1,587 @@
...
+    // Map each criteria column to a database column index (or -1 if no match)
+    const criteriaColumnMapping: number[] = criteriaHeaders.map(criteriaHeader => {
+      if (typeof criteriaHeader !== 'string') {
+        return -1
+      }
+      const lowerHeader = criteriaHeader.toLowerCase()
+      return dbHeaders.findIndex(
+        dbHeader => typeof dbHeader === 'string' && dbHeader.toLowerCase() === lowerHeader
+      )
+    })
```
[the remark this answers is not in the document]
Comment (marcin-kordas-hoc) [reply src/interpreter/plugin/DatabasePlugin.ts:522]: Fixed in commit `74c439725` — added an explicit pre-pass over `criteriaHeaders` that returns a `CellError` immediately if any header cell contains one. Aligns with the data-cell propagation at lines 535-537 and matches Excel behavior. Test added in private repo `test/hyperformula-tests/unit/interpreter/function-database.spec.ts` (`propagates a CellError from a criteria header cell instead of silently skipping the column`).

**Review by sequba — APPROVED** [review]

**Review by sequba — APPROVED** [review]
